pkg/sql/catalog/typedesc/virtual_table_record_type.go

```diff
@@ -0,0 +1,316 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package typedesc
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/errors"
+)
+
+// VirtualTableRecordType is an implementation of catalog.TypeDescriptor that
+// represents a record type for a particular table: meaning, the composite type
+// that contains, in order, all of the visible columns for the table.
+type VirtualTableRecordType struct {
+	// desc is the TableDescriptor that this virtual record type is created from.
+	// TODO(jordan): does it make more sense to eagerly pop out all the fields
+	// that we care about from this backing TableDescriptor into the struct,
+	// rather than have all of the methods delegate?
```

Comment:
My personal answer to that question would be: no, keep on delegating. From the point of view of the lifecycle of a record type descriptor, it's clear that its underlying the table descriptor will never change for the duration of its existence. Plus they're so tightly coupled.